### ISSUE TITLE
Functional properties don't get inferred with circular type parameter constraints like `T extends M<T>`

### ISSUE THREAD

**Issue by devanshj** [microsoft/TypeScript#40439]
**TypeScript Version:**
4.0.2

**Search Terms:**
Circular type parameter constraint, functional property inference, function parameter inference, function argument inference

**Code:**
```typescript
declare const m: <T extends M<T>>(m: T) => T
type M<Self, K = Exclude<keyof Self, "k" | "t">> =
  { a?: number
  , b?: number
  , c?: number
  , d?: number
  , k?: K
  , t?: (k: K) => void
  }

// :)
// Case 1
m({
  a: 1,
  b: 2,
  k: "a"
})

// :)
// Case 2
m({
  a: 1,
  b: "x", // expected error
  k: "c" // expected error
})

// :(
// Case 3
m({ // type parameter becomes `unknown` (probably because t's first parameter is `any` initially kinda?)
  a: 1,
  b: 2,
  k: "a", 
  t: k => {} // k is inferred as `never` (instead of `"a" | "b"`)
})

// :|
// Case 4
m({
  a: 1,
  b: 2,
  k: "a", 
  t: (k: "a" | "b") => {} // have to explicitly type `k` which could have been inferred :(
})
```

**Expected behavior:**
The parameter of `t` should be inferred as `"a" | "b"`. As property `k` already gets inferred in Case 1 & 2, so why not the `k` parameter of `t`

**Actual behavior:**
The parameter of `t` is inferred as `never` and `T` in inferred as `unknown`. I assume first `T` gets resolved to `unknown` as `t` is initially inferred as `(k: any) => void` which does not satisfy the constraint of `m`.

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3iIEp4BeRjgFAYAngAcEDAMogIAMwA08ANK94AUTKQArqAIBrEMKSz40uYoBEei-AA+8CxgvNeA+PADe8KAH58KLWwAIxAYN3hFIL94AODQ8MUwaNiQsPdFYGTA1IT4PWilXIxolj18JW4+eAA3JABLYHCAXwEBAHo2+FxOds6AYSg0BABGAWwWD3CofGH5cKD8ACY59zKHKAsBJp7erp6O+AGh+EWxiamZlfgFhzILRQPycTBKYFIYGCQ0vPwLMBtHmRnq93p8wttWgdcCxdkcEABmM5eA4icTwUSwKDYECUGDXcBIbG0AAGWhQehQSAA7ihifAWKJPkEoEEIMJ8WAoFpjhgAOS0WR1GCYdGY7G4+B1ElQFDCOl1FB1DB1KAQNl5BXAXw9dzTeCzeZLK5rCwbRThDD4PS8fgeJrwA7WqWSlCyUJwN6DeDElAgaqhOksBWYEBQN7Gb2mmz2CxBCzEnoQ3a4WywwYIAAsSIu+quN2W4RNZvgFvwpV+GzsDjjlVt9oOAAsoP74BgkKQgRA6mAlerUQhiXo6VSG92G4gkFoIG8my2QlQXW6PiA3tCtpwgA)

**Comment by devanshj** [comment]
@jcalz if you could help in anyway it would be great! 🙏🏻

**Comment by devanshj** [comment]
Workaround:
```diff
- declare const m: <T extends M<T>>(m: T) => T
+ declare const m: <T extends M<T>>(m: InferObject<T>) => T

+  
+  type InferObject<T> =
+    {	[K in keyof T]:
+        IsPlainObject<T[K]> extends true
+          ? InferObject<T[K]>
+          : T[K]
+    }
+  
+  type IsPlainObject<T> =
+    T extends object
+      ? T extends (...a: any[]) => any
+        ? false
+        : true
+      : false
```
[\* tears of joy *](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3gEkUAzEGAeQBGAK3AYmASngBeRvCIAoBRgCeABwRdeAkWKYyF8eAG8AkAG0A0vACWKeAGsQKpN3kBdXIaM+OaAArQdkKiYOJEVu5y5JTUtBgwAK4g3j4+APycPHwhehGWUalp8Pj57t4AvkqqGpwBQSi5YfrS3iQxVDTwSLphRZntFJ20LAB041D4UCgq5u5SsvDTKkVGmdxQEGgpxSXwCclF+BtbKcrqCAwAyiAQ3AA08NbS8ACiZJCJoAROLm43d0eACIHED4AAfeBAjBA5gGIzGJbpfAoRLYQR8byPQTI+Co9GYoyPMC4-EYmBY+DAUlo8mUhy4yyUjC4lgOfCWBZyABuSBswEqSgA9EKSgAqeAKbAsYzeSbwACM928gnwACZlUZ2VCoEDHt4MPgHDI5MYKvARY5bLQ7No4MAlrQAAZA3UQqGCIFOhQVCRAA)

I'll keep this issue open to know if this is the "official" way to deal with it or not.

**Comment by devanshj** [comment]
FWIW the #44700 variation makes this seem actually a bug

**Comment by RyanCavanaugh** [comment]
Pasting in that repro for reference
```ts
declare const m: <T extends M<T>>(m: T) => T

type M<Self, V = Prop<Self, "v">> =
  { v: V
  , t?: (v: V) => void
  }

type Prop<T, K> = K extends keyof T ? T[K] : "error"

// expected : M<{ v: number, t?: (v: number) => void }>
// actual   : M<unknown>
m({
  v: 1,
  t: k => {}
})

// expected : M<{ v: number }>
// actual   : M<{ v: number }>
m({
  v: 1
})
```

**Comment by RyanCavanaugh** [comment]
What you have here is equivalent to the much simpler, and also working,
```ts
declare const m: <T>(m: M<T>) => T

type M<T> = {
  v: T
  t?: (v: T) => void
}

// expected : M<{ v: number, t?: (v: number) => void }>
// actual   : M<unknown>
m({
  v: 1,
  t: k => {}
})

// expected : M<{ v: number }>
// actual   : M<{ v: number }>
m({
  v: 1
})
```
If I had to wager about the confusion, it's that the default for `V` is not an existential, but rather an *actual default* that gets instantiated in-place when not specified.

If you have a description of `m` that can't be typed straightforwardly, I'd be happy to take another look

**Comment by devanshj** [comment]
First off I was utterly confused because in the code snippet you wrote `// actual   : M<unknown>` instead of `// actual : M<{ v: number, t?: (v: number) => void }>` xD It's only when I read "and also working" and tried it in the playground, I realized it indeed works.

> If you have a description of m that can't be typed straightforwardly, I'd be happy to take another look

First to give you some context - The thing is I'm using these `T extends M<T>` style types (I call them "Self Reconstructing" types) to do a variety of cool things (like show custom errors) and mainly strongly type [xstate](https://github.com/davidkpiano/xstate) in my project called [txstate](https://github.com/devanshj/txstate) (which will hopefully be merged with xstate itself). Some of these style of types have already landed xstate like [here](https://github.com/davidkpiano/xstate/pull/1999).

Though to give you a concrete problem and as minimal as I offer right now, you can refer [this PR](https://github.com/cassiozen/useStateMachine/pull/36) (to another xstate-like library called useStateMachine) and precisely [this test case](https://github.com/devanshj/useStateMachine/blob/6a639c2b517e3208e33ab407fa89e804830d202a/test-d/useStateMachine.test-d.ts#L632-L649). To make it work you'd have to add a phantom property `_: null` beside the effect property like so...
```ts
useStateMachine({
  initial: "a",
  states: {
    a: {
      _: null, // Remove this and the inference stops working
      effect: parameter => {}
    }
  }
})
```

[This same is the problem with xstate too](https://github.com/devanshj/txstate/blob/d195b5170db31ed83c13109ecb8c3698a5e18a48/src/MachineDefinition/MachineDefintion.Entry.spec.ts#L112). So it'd be really cool if you can take another look and see if TypeScript can do a better job of inference. Thanks a lot for your time! :)

**TLDR**: Open [this test case](https://github.com/devanshj/useStateMachine/blob/6a639c2b517e3208e33ab407fa89e804830d202a/test-d/useStateMachine.test-d.ts#L632-L649). Remove `@ts-ignore`. Verify is you can see errors. Add a phantom `_: null` as a sibling property to `effect` (like shown above). Verify if the error is gone. State machine libraries like xstate and useStateMachine can be only strongly typed with this "non straightforward" way.

cc @davidkpiano @Andarist @cassiozen

**Comment by RyanCavanaugh** [comment]
Sorry for leaving in the bad comments, good reminder for me for next time 😅

**Comment by devanshj** [comment]
Still finding a real world repro for #44700 variation, though I have one for the original problem. Notice we need `InferNarrowestObject<D>` workaround (same as the `Identity<T>` workaround above) to make it work. There isn't a straightforward to strongly type this afaict (if one is even successful these libraries have sophisticated requirements than this snippet)

This is the workaround to be clear
```diff
- declare const useStateMachine: <D extends Machine<D>>(definition: D) => void
+ declare const useStateMachine: <D extends Machine<D>>(definition: InferNarrowestObject<D>) => void
```

[Playground](https://www.typescriptlang.org/play?#code/K4ZwpgygLghlYFkYGMAWBLAdmAFAbwCgACIrdKdGAGwC4iAiGegGmKOQHtN4APKOvEQBmHDnQCMRAL6sSIWPBAC2JGMpIaiXdZo0ANOvQBG9FRqlmZZozs1ghQsMn5EADkQC8APiKFdGgHoAogABKBAAWjAeVycoKIAnBI4Eohxo2OcwABMiMCSU4ULFCkwAcwBKM01XADpObmioWpEOTw8PBiYiasDgsMiMuMTk1PSYuJy8gtSRVJKsSt6SOrAANzBuWqgAT1j2zvoeU38LTTPpAikqgmynKhgEsHYueSJQSAVEFAxsOgAeAAieT4m2yICISDQWDAQK8XhwdyEZAo2iIAElMI4EgA5R7JADuYHkAHkjAArOJwiqeHxrDjobIEAi7fZQ36wiBgKhCHydNiCFGUWhEADWYB2HCERAA4mAoP8uTzmAx5HBifQvGwVQ1eC45QqlUIVfRdU1NdqiGrFHQzIIANoQUiYMUSqWy+WK7nG1VfECagC6NGWRGg6pxHDuXuVHsN3pV9vo1o1KogAfhljYFhZe2eYfgEaj7JhqfjRBJLs6BujPvoXE1fIFWkwAH5g5oHQBRZ2uyXSitByE-GH5sAANWowFhxewWvOlvsjmcbd6OFcjxgAFt5fl2-5fC9Gnw6DOwABhLh6-6nuf7lXrTYuU+djbca-D2chi6aGneIj0xks2ZVlnlPC8jwVG9PFjd9oWwE0zT4C0QKHODPnVCcqCnWCOT5Xt3WrU8TWTf0tRQ59X0gj8wEbEhBBAvdNDJSlnEw7CHSdLB8OlQjqOIv1A0Y-dmLiNjYS7HtxT7GCiKIRMSJYUMTXrdMgxDfxeLQhMkwE1NlMwRTO3TdTpFvc5b2zFDqwAFRVAAFFUADFoOATAkRhbJaKIOyQXgNyISk91rLMFsiGs+07IDIg9x8powQhe0opbMxrN8+L3jc+wPKIULnLoVKYrS-y5KwbEiAAaXEFVahq0r8gqhIkrMSqivBbiwpy3obPtSqAxVcqEicnwhLythsA2BIc32ETWMnWFrLw8LAulayA2ZKbnkxbE8QKIl5H+BbPDYVK4uKmBMB2EK0g6062pwGrajUIhzp2RLfx8F6co6oTNBO0Fio4Ck4i+rb8h2wliSgGaFUOn6NGC-wqk0OhxvyYDcwxLEwfxDg9qhoHnAOrzBCeGBsi4Kgdjk8rJLdFbB1B3Ecbxg6evTaQgA)
```ts
useStateMachine({
  initial: "a",
  context: { foo: 1 },
  states: {
    a: {
      on: {
        X: "b"
      }
    },
    b: {
      effect: p => {
        // @ts-expect-error (expected error for testing)
        p.context.foo === "a" 
        // @ts-expect-error (expected error for testing)
        p.event.type === "x"
      }
    }
  }
})

declare const useStateMachine: <D extends Machine<D>>(definition: InferNarrowestObject<D>) => void

type Machine<Self> = 
  { initial: keyof Get<Self, "states">
  , context: Get<Self, "context">
  , states: 
    { [S in keyof Get<Self, "states">]:
        StateNode<Self, Get<Self, ["states", S]>>
    }
  }

type StateNode<Machine, Self, On = Get<Self, "on">> =
  { on?:
    { [E in keyof On]: MachineStateValue<Machine>
    }
  , effect?:
      (parameter:
        { context: MachineContext<Machine>
        , event: MachineEvent<Machine>
        }
      ) => void
  }

type MachineContext<Machine> = Get<Machine, "context">
type MachineStateValue<Machine> = keyof Get<Machine, "states">
type MachineEvent<Machine> =
  { type:
      ObjectValue<{ [S in keyof Get<Machine, "states">]:
          ObjectValue<{ [E in keyof Get<Machine, ["states", S, "on"]>]:
              Get<Machine, ["states", S, "on", E]>
          }>
      }>
  }

type Get<T, P, F = undefined> =
  P extends keyof T
    ? T[P] :
  P extends [] ?
    T extends undefined ? F : T :
  P extends [infer K1, ...infer Kr] ?
    K1 extends keyof T ?
      Get<T[K1], Kr, F> :
    F :
  never
type ObjectValue<T> = T[keyof T]


type InferNarrowest<T> =
  T extends any
    ? ( T extends (...a: any[]) => any ? T :
        T extends object ? InferNarrowestObject<T> :
        T
      )
    : never

type InferNarrowestObject<T> =
  { readonly [K in keyof T]: InferNarrowest<T[K]> }
```

**Comment by devanshj** [comment]
Repro for the phantom "_" bug. [Playground](https://www.typescriptlang.org/play?#code/PTAEEkDsEsBdoIYBtTQM5oK4FMBQBjAJ2wVmwFkF8ALaSbARgAoBvXUVGeZALlACIE-ADTtQ+APaQyAD1h8WoAGYSJfBqAC+ojmlilsaBWI4JjHC6GzTCATz4AHUAF4AfKEUhQT9KASRbUB4mEwtNUPCOcM0ASlxcLwB3OGoISCVsQgA5BEJCCUTDWAB5ACMAK2x8WFBEiUIAa1yJTEgAEwJiA0oaOmwAJlYxOjhEJD5BETFJaWw5BWVVdS0dUD0DIw9Qsy3LDmtYO0cXd08wHzQPRbVQSEwAW1LMrVCo1csAfT47pCRhUC8sGoElq2Fu2GwbVAQN8Dmo-lgEnu3nyDkysFsADpsRExNE4gkwNjMW0JIZIAByGp1Rq1FItGr8D78TokMg9Wj0QZsDgjbjjARCVYzWTyK4qG4abRidZkTY8iw7BWWA5HbwnXZ7SJhPG4WLxNpVJC5MEzPTiLrsqicxh8AA8ABErHJrG1Lhy+o7XK4mIalHzoFI+A6YhqAG4SaAdQ34Y3EcRSc1ENkUa19fr2p1zMjtd1p+hen1+gNBtIZbK5fKFPRlSrVL2htygCNR3AYtGgD0FgDK2CQSnczlAYkUAd4oAa2FsEiUoAACqi7b3+-9+LLDPxXGJ-iLs3wFxIHEu+0pV7u5Jvt2t9HKeKFFABtbucCdTmfzxfL08CddoTcAXTvPYLEUVVbAAfj4JhzzFA8jy-M8pFFTdG3cFsOmAjh-g+SDbkwX5Xh1SxtXCNtbA7KByxyPICiKO0ABVBzEejnRzN0-ACUJwNAJhQBY7NXUuJhiR2fxbAff9UI4wJuJYoDMP4l1c1ACQKiqGpuMozJqKrIpa3Uhj3Hk4D6MIuJLG+bAw0yeJ2zBLSKxo6sSjU+tGJcEdQC6UlICQQIHwAaRfSdp1nejALLbTK1ovQGMC-93FIuyP0PBj-jnJiODnVjBNfUK+NAWSHznf8gnBazCCAA). I know you can [type this in a straight forward way](https://www.typescriptlang.org/play?#code/MYJwpghgLmCyHABYEsB2YAUBvAUAAjzWSmQgBsAuPAIgmoBp89gB7VGADyiqzwDMWLKgEY8AX0YEAzlGhgpPJgQiKCavGHYgAnlQAOeALwA+PLnUExS8UytiAlDhwATMMDIRwzNjObg58EhoYFQAPADKGlyazlJ4MiBoAOb0eADCxhiufEQkbFSBKOgRqRn2RqYAbizIzjhQ2npgeIXBEVEwqLHxUImoKemmhnhMvLmklHgAciwAkqh8YCARxkyprOxgXFRpaz1yCta8ANoA+oSoeOEAuhTW6ryavdoA-FQYG5zc03MLS6FlCp4aq1e5qKzqCE2eqNZozeaLZYAFSGeGOSOu6I6MTiEFQ2jwLzwAAY8FR0JUltcgA) but it's besides the point for the above reasons. It's only the case here because this is a minimal repro, in reality the requirements are much more.

```ts
// Initial issue
createMachine1({
  initial: "a",
  context: { foo: 1 },
  states: {
    a: {
      entry: p => { // p is any :(
      }
    }
  }
})

// with InferNarrowestObject workaround
createMachine2({
  initial: "a",
  context: { foo: 1 },
  states: {
    a: {
      entry: p => { // p is { foo: number }
      },
      _: null, // tho we need this phantom property...
    }
  }
})

// ...doesn't work without "_"
createMachine2({
  initial: "a",
  context: { foo: 1 },
  states: {
    a: {
      entry: p => {
      }
    }
  }
})

declare const createMachine1: <D extends Machine<D>>(definition: D) => void
declare const createMachine2: <D extends Machine<D>>(definition: InferNarrowestObject<D>) => void
type Machine<Self> = 
  { initial: keyof Prop<Self, "states">
  , context: Prop<Self, "context">
  , states:
    { [S in keyof Prop<Self, "states">]:
        { entry?: (context: Prop<Self, "context">) => void
        , _?: null
        }
    }
  }

type InferNarrowest<T> =
  T extends any
    ? ( T extends (...a: any[]) => any ? T :
        T extends object ? InferNarrowestObject<T> :
        T
      )
    : never

type InferNarrowestObject<T> =
  { readonly [K in keyof T]: InferNarrowest<T[K]> }

type Prop<T, P> =
  P extends keyof T ? T[P] : never
```

**Comment by devanshj** [comment]
Here's something quite real-world that can't be compiled in a "straight forward way". [Playground][1]. This is as straight forward as I can manage. 

```ts
// does not compile
createMachine({
  initial: "a",
  states: {
    a: {
      entry: c => {}
    },
    b: {}
  }
})

// remove entry and it compiles
createMachine({
  initial: "a",
  states: {
    a: {
    },
    b: {}
  }
})

declare const createMachine:
  < Definition extends
      { initial: keyof Definition["states"]
      , states:
          { [K in keyof Definition["states"]]:
              { entry?: (k: K) => void
              }
          }
      }
  >
    (definition: Definition) =>
      { definition: Definition
      }
```

[1]: https://www.typescriptlang.org/play?#code/PTAEBMHsFMGdQHaQC6gMaQLYAcCWAbaAKDQCdoBDZaAWQrQAtcFoAKAbyNFGd2Vwr4AXKABEFUQBouoWMipwRnbtwpKZK0NATJSATxFpQAXgB8odgF8NoS9M0AjJde7XLASiJEQocpkgAbtBaOvqgFAjgPKgYOARwJOQKdIzMbMo8CHwCwmIS9rLy1LDqmmoWNnY2ThYutkQeXuDQaPgU5OiQCHLoSdQpTCxCMgA8oAAi0ABmvPxdWgAe1JGwNtzsmdmCIgDW0HqQUxPTs7hdANqicgqwogC6a6CShTfDmu8WoOcA0pmgewcjpMZlk5ghLtdivc7m8PnCNtpdHoAPwiVg7ETfdwmcwBSC4cCPOH1YmuR51Uw2VjNEHZLoiYGnLrYsyPDY0pkIBknUFnBDkoA

**Comment by devanshj** [comment]
@RyanCavanaugh what are your thoughts on this? Adding a functional property toppling the inference is fishy if not buggy. Especially when parameters are explicitly typed as what were expected to be inferred, it compiles.

And how likely/unlikely is that future TypeScript versions would break [this workaround](#issuecomment-689729651) or in other words how reliable is it?

I'm asking especially because I'll be using this workaround in [`@cassiozen/usestatemachine`](https://github.com/cassiozen/useStateMachine)<sup>1</sup> so want to let users know how risky it is. And I'll also provide a workaround-free degraded version of types so vigilant user could use that.

<sup>1</sup> which is decently popular to say the least and with my types it's literally going to be the best state machine library out there in terms of types :P so I'm expecting a lot of people are going to use it and I want to let them know how risky it is or is not.

**Comment by RyanCavanaugh** [comment]
@devanshj unless I'm missing an existing trick we could re-apply here, the example above is squarely in the realm of "need a unification-based algorithm". The difficulty here lies in the self-same mapping between the keys and callback values of `states`.

This definition is *possibly* better depending on your use cases?
```ts
type SelfStates<Keys extends string> = {[K in Keys]: { entry?: (k: K) => void } }
type Definition<Initial, States> = { initial: Initial, states: States };
declare function createMachine2<StateNames extends keyof States, States extends SelfStates<StateNames & string>>(def: Definition<StateNames, States>): Definition<StateNames, States>;
```

**Comment by devanshj** [comment]
> unless I'm missing an existing trick we could re-apply here, the example above is squarely in the realm of "need a unification-based algorithm". The difficulty here lies in the self-same mapping between the keys and callback values of states

I'm a total novice but I think that's right. But more precisely I think what is happening is (perhaps that's what you meant) is that `K` in both what is meant to be "passed to the `createMachine`" AND what is meant to "come from `createMachine`" via the callback parameter, so because of the heuristic Anders mentioned [here](https://github.com/microsoft/TypeScript/issues/44999#issuecomment-883531098) the compiler decides to give callback `K` the priority and resolves it to `any` -- same thing happens in every variant of this issue. So it being a key is not super relevant as long as it's something that "goes" to the parent generic function hence even though keys are not involved [here](#issuecomment-866330460) the issue still remains.

So I think providing a `LowInfer` which I suggested [here](https://github.com/microsoft/TypeScript/issues/44999#issuecomment-883717710) would be simpler than a unification algorithm, though I'm not sure if it would simpler nor I'm sure if it'll solve the issue.

> This definition is possibly better depending on your use cases?

It does seem to behave better (I think I did try this) beside the fact if you have more than one states [it doesn't work][1] but can be solved if you [assert `initial` to be union of states][2] (I did try `NoInfer<Initial>` but apparently [doesn't work][3]).

But ultimately it still doesn't cover my use-case of having the access to the whole tree, it turns out the my last repro that you just solved is too minimal. Let me try to increase it's requirement to the original tad bit...

```ts
// compiles
createMachine({
  initial: "a",
  states: {
    a: {
      onNext: "#someId"
    },
    b: {
      id: "someId" as const
    }
  }
})

// add entry and it doesn't compile
createMachine({
  initial: "a",
  states: {
    a: {
      onNext: "#someId",
      entry: k => {} // because k doesn't get inferred
    },
    b: {
      id: "someId" as const
    }
  }
})

// annotate entry that could have been inferred and it compiles
createMachine({
  initial: "a",
  states: {
    a: {
      onNext: "#someId",
      entry: (k: "a") => {}
    },
    b: {
      id: "someId" as const
    }
  }
})

declare const createMachine:
  < Definition extends
      { initial: keyof Definition["states"]
      , states:
          { [K in keyof Definition["states"]]:
              { id?: string
              , onNext?: 
                  | keyof Definition["states"]
                  | { [S in keyof Definition["states"]]:
                        Definition["states"][S] extends { id: infer Id } ? `#${Id & string}` : never
                    }[keyof Definition["states"]]
              , entry?: (k: K) => void
              }
          }
      }
  >
    (definition: Definition) =>
      { definition: Definition
      }
```

I can take this a step further and [not require the user to write `as const`][4] but let me keep it as minimal as possible :P

[1]: https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhA2AzGwCGwIGcA8BpCImUEAHhgHYAmRmwATgJbkDmAfFALxQDeA2rlCZR8hALoAuHsXL0QAfkkAKANaTcASk7sAbgHsGlKAF9jAKFCQoAEQiImDYA13lsASXIOGqeABpYaDEx2Lm5BD0dvSXdPbz9adCxJFASiIwBuU0oIAGN4VDpoRABXcmzHZyhsgoSAWVRsgAsmCAAmbGSMADlUAFssYjIIKiJlAl1EfxS-Dv7SCmpYBGQArHaV7r6iADIoWkYWVlZFLMRJGztwpxcZjaxplaD1M9t7cuv13rvJwNYM0yqILV6k1yK1FNxTFAwjF4JIAESoOE+SG7B6SCFQqGodEozHSWSSZRaHimXGYoxkozIvEAIxxeKhQwJUCJHHYGIZULJUIpeN5ZiM6iAA

[2]: https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhA2AzGwCGwIGcA8BpCImUEAHhgHYAmRmwATgJbkDmAfFALxQDeA2rlCZR8hALoAuHsXL0QAfkkAKANaTcASk7sAbgHsGlKAF9jAKFCQoAEQiImDYA13lsASXIOGqeABpYaDEx2Lm5BD0dvSXdPbz9adCxJFASiIwBuU0oIAGN4VDpoRABXcmzHZyhsgoSAWVRsgAsmCAAmbGSMADlUAFssYjIIKiJlAl1EfxS-Dv7SCmpYBGQArHaV7r6iADIoWkYWVlZFLMRJGztwpxcZjaxplaD1M9t7cuv13rvJwNYM0yqILV6k1yK1FNxTFAwjF4JIAESoOFQVBEBFIgA+UDhACM4T5IbsHpIIVCoahiQTSdJZJJlFoeKZKaSjEyjPiqdiKVSoUMaVA6Rx2BCmVCWVSxWYjOogA

[3]: https://www.typescriptlang.org/play?#code/C4TwDgpgBAyhA2AzGwCGwIGcA8BpCImUEAHhgHYAmRmwATgJbkDmAfFALxQDeA2rlCZR8hALoAuHsXL0QAfkkAKANaTcASk7sAbgHsGlKAF9jAKFCQoAEQiImDYA13lsASXIOGqeABpYaDEx2Lm5BD0dvSQA5XXdECDo3cK94Vj9adCxJFEyiIwBuU0oIAGN4VDpoRABXchLHZygSysyAWVQSgAsmCAAmbByMKNQAWyxiMggqImUCXUR-XL9B8dIKalgEZACsAZ3hsaIAMihaRhZWVkVixEkbO2TnPcyDrGWdoPU723sGlxXXph3rlWIVzOBoDE4glsAAVYJQXiw0RIibrIiocggKByKAABigknIEG0CVEplMzQgbQ63WJvUU3FMUDCnkiUAARKgOT5macPpImSyWahBXzhdJZJJlFoeBSJSyjOLjLyJQAjMUKyV0EDS2VM5WK5VKo1GdRAA

[4]: https://www.typescriptlang.org/play?#code/PTAEGMHsFsAcEsA2BTAzgKHAJ2QQwC7ICyu4AFvAHbIAUA3uqKFfPvLogFygBEuPAGkahU+Amm4MmTXJOHTQkSgDlkAD3zceAYlQxkASQAmPeaAC+QhQCM5CpvCNa90QybPnhn8wEp06EFBcIyNQZEp8LABPIMpQ1lAjSDRKAHJ8CBgEFEwccRJyKlopZkpWdi5efisRMUJUO2lZUBKFJVUNLV19Y0EzJnDIqO4Aa1AAXgA+FvNQQOtkcFwAV1RkUDGklPTQAHNkDKoAM2QsHCMPGqZbFv7mJ14XN1MFTyZvPwCwXEpKSDr1oNoqB8GQCJllohQmCAG7rBbhUonM7IUI-eIZKBwJBoXJ4QgFCjUejCFhsDhaarCUTiBq3BTNVrSdrqTS8bquXpXaRA4agGgjSk8HwTaZ0N7SSxmG5MhwPHhPXoeLzoXz+IyLRC4HCZSiiCB5AmkInITjCAA8oAAIsgjmT4EowhpwkYMPYWqVyhSNsgopAjtbbfalABtBUA1A8AC6dwEtVpZvdCjooBDAGlSj6-QGbXaymxQ+HadGo4mk+6U44APzcXPByhhmn1aPpqNhxzRp2EOKoJGnUDGUBVgeUZEAZUiVF2ABlWKcOObjNNuNQ4Vg7uW4yyNDXQBvy9IAD5Z-2BvPlQtNtDR-cH0DHlMhseZka+091-MOhtF5tR0u3u97A-C9vyvSM2zHKMuxdXtKweY5+0HWZhwAA20AASOhBwAMlqLAp3MFDQBXZA1wAu9zBDV9szPetGwjEtyOkONeV3AVuDTEUplAGFIEcJj3gAiVXmESYzBoDVzwLShayDT8lC4sSK0SOSQNkqSvzuTx0HwKJYHWAxR1OCd8MoGc5ywBcABVpnGYQrOgns8KnIdQAc7hRFM3Z0CAA

**Comment by Andarist** [comment]
This suffers from a similar problem as https://github.com/microsoft/TypeScript/issues/48798 . It is somewhat different though because this one here doesn't contain any reverse mapped types.

```ts
m({ // type parameter becomes `unknown`
  a: 1,
  b: 2,
  k: "a", 
  t: k => {} // k is inferred as `never` (instead of `"a" | "b"`)
})
```

1. this arrow function is "context-sensitive" because it contains unannotated params
2. so `checkFunctionExpressionOrObjectLiteralMethod` returns `anyFunctionType` for it and that has `ObjectFlags.NonInferrableType` on it
3. this flag is one of the `PropagatingFlags` so the parent object (the argument itself) becomes non-inferrable
4. the signature gets inferred as `const m: <M<unknown, never>>(m: M<unknown, never>) => M<unknown, never>` because of that
